### ISSUE TITLE
Function by method bodies should not be allowed to contain expect statements

### ISSUE THREAD

**Issue by robin-aws** [dafny-lang/dafny#1376]
Similar to #1346, which is about tracking the side-effects of `print`, in particular in the context of function by methods, but arguably more concerning.

```dafny
function method GoodFunction(n: nat): nat {
  expect n > 0; // Error: expect statement is not allowed in this context...
  0 / n + n
}

// Verifies and compiles without errors
function BadFunction(n: nat): nat {
  n
} by method {
  expect n > 0;
  return 0 / n + n;
}
```

This creates a function whose specification-time behavior does not completely align with its compiled behavior: `BadFunction(0)` is equal to `0` and yet halts at run time.

In retrospect, I would argue that `expect` is tightly coupled to termination proofs: a failed expect statement causes a method to fail to terminate just as infinite recursion would. I propose forbidding `expect` statements in any method that does not have `decreases *`, which would would then forbid its use in function by methods since they cannot skip termination proofs with that syntax. This would break backwards compatibility but could be reasonable, given `expect` is only used in test code to the best of my knowledge.

**Comment by hmijail** [comment]
How would this apply to the documentation's example of using `expect` to check an `:extern` result?

**Comment by robin-aws** [comment]
It just implies the method would need to declare `decreases *`, which is allowed since the example is a `method` and not a `function method`. That's arguably a more accurate declaration of the `:extern` signature anyway, since there is no guarantee external code will terminate.

It would be useful in the future for Dafny to still support termination proofs of methods that call other methods with `decreases *` somehow, but there isn't a concrete proposal for that yet.

**Comment by keyboardDrummer** [comment]
> I don't think we should make it possible for user code to recover from halting though. If some feature higher in the "stack" was able to recover and continue execution, there would be no guarantee that the post-conditions of all the methods on the stack were satisfied, so you would be in major danger of runtime errors that should have been proven impossible.

In case you call a method that halts then I think you wouldn't do anything with the post-condition. Could you not compute a replacement post-condition based on the modifies clause that implements the maximum disturbance the method could have?

To be clear, the reason I mentioned a recovery mechanism is that otherwise it's not clear to be why you would want to track whether code can halt or not.

**Comment by robin-aws** [comment]
Having mulled this over for a while, I've arrived at a concrete alternate proposal.

`expect` is a way of knowingly introducing unsoundness into your verified code. It's intended to let you avoid proving something you believe to always be true, without writing a bunch of error handling you expect to be dead code. By extension this makes it well-suited to writing tests, but also to interacting with external code where if an assumption is violated there's no reasonable way to continue execution.

I assert that by default `expect` should not be allowed anywhere in a Dafny codebase unless you opt into that unsoundness. You should be able to pass a `/allowExpect` flag if you want to use this feature. The typical default layout for a Dafny codebase would be to only pass this flag when building/verifying/testing the `test` source directory but NOT the `src` directory by default. 

As an aside, I've wanted to find a way for Dafny to support more "incremental verification", so that people can get started and make progress on Dafny code quicker, while tracking the fact that they still have work to do to prove correctness. This would mirror the benefits of gradual typing. I've arrived at the conclusion that `expect` is ideal for that: you can count all the expects in your code as a metric to drive down over time, and likely set up the mechanism to forbid code built with `/allowExpect` from ever making it to production.

**Comment by RustanLeino** [comment]
I agree with this view of `expect`. Having an `/allowExpect` flag without which _the compiler_ (not the resolver or verifier) complains upon encountering an `expect` statement is also fine with me.

To make sure I understand the alternate proposal, @robin-aws , you're suggesting that we not have any other finer-gain `expect` checking, right? That is, either you opt in and use the `/allowExpect` flag and that means you can use `expect` anywhere, or you don't use the `/allowExpect` flag and then the compiler will refuse to output any code that has an `expect` statement. Is that right? If so, that sounds fine with me.

As a general design point (now addressing @keyboardDrummer 's points and questions), I think that it seems reasonable for a verified language to offer some kind of control over each possible effect in the language. Since `expect` statements can bring about a "halting effect", then having some way to control it seems reasonable to me. The `{:expect}` attributes I proposed above are perhaps overkill, but `/allowExpect` would also satisfy me.
